oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOpaqueTokenIntrospector.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.spring.boot.oauth;
+
+import com.nimbusds.oauth2.sdk.TokenIntrospectionResponse;
+import com.nimbusds.oauth2.sdk.TokenIntrospectionSuccessResponse;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import com.nimbusds.oauth2.sdk.id.Audience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.MediaType;
+import org.springframework.http.RequestEntity;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.core.DefaultOAuth2AuthenticatedPrincipal;
+import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal;
+import org.springframework.security.oauth2.server.resource.introspection.BadOpaqueTokenException;
+import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionException;
+import org.springframework.security.oauth2.server.resource.introspection.OpaqueTokenIntrospector;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestOperations;
+
+import java.net.URI;
+import java.net.URL;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Okta's implementation of Spring security's interface {@link OpaqueTokenIntrospector} based on
+ * {@link org.springframework.security.oauth2.server.resource.introspection.NimbusOpaqueTokenIntrospector}.
+ */
+public class OktaOpaqueTokenIntrospector implements OpaqueTokenIntrospector {
```

Comment:
There seems to be a lot of logic in this class.  Why not use `NimbusOpaqueTokenIntrospector` instead?

Reply:
This is because, we need to customise the request body and headers to send additional params to Okta's introspection endpoint. The `NimbusOpaqueTokenIntrospector` just sends the token to the introspection endpoint and this wont suffice. Hence the customization.

Reply:
I'm not sure I'm following.
You can just set the auth headers in the `RestTemplate`

https://github.com/spring-projects/spring-security/blob/bf067d679fb5990d6b3d31cee292af171f8a5014/oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/NimbusOpaqueTokenIntrospector.java#L82-L83

The client-id and secret are not needed when you configure the RestTemplate directly

Reply:
got it, updated.